File: WINGS_AI_Bootcamp/Python/PasscodeDerivation_ProjectEuler79.py
from collections import defaultdict, deque

def derive_passcode(attempts):
    graph = defaultdict(set)
    indegree = defaultdict(int)
    digits = set()

    for attempt in attempts:
        digits.update(attempt)
        for i in range(len(attempt) - 1):
            if attempt[i+1] not in graph[attempt[i]]:
                graph[attempt[i]].add(attempt[i+1])
                indegree[attempt[i+1]] += 1

    queue = deque([d for d in digits if indegree[d] == 0])
    result = ''

    while queue:
        node = min(queue)
        queue.remove(node)
        result += node
        for neighbor in graph[node]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                queue.append(neighbor)
    return result

attempts = [
    "319", "680", "180", "690", "129", "620", "762", "689",
    "762", "318", "368", "710", "720", "710", "629", "168",
    "160", "689", "716", "731", "736", "729", "316", "729",
    "729", "710", "769", "290", "719", "680", "318", "389",
    "162", "289", "162", "718", "729", "319", "790", "680",
    "890", "362", "319", "760", "316", "729", "380", "319",
    "728", "716"
]

print("Derived passcode:", derive_passcode(attempts))

File: WINGS_AI_Bootcamp/Elixir/PasscodeDerivation_ProjectEuler79.ex
defmodule Euler79 do
  def derive_passcode(attempts) do
    edges = Enum.reduce(attempts, %{}, fn attempt, acc ->
      [a, b, c] = String.graphemes(attempt)
      acc
      |> Map.update(a, MapSet.new([b]), &MapSet.put(&1, b))
      |> Map.update(b, MapSet.new([c]), &MapSet.put(&1, c))
      |> Map.put_new(c, MapSet.new())
    end)

    all_nodes = Map.keys(edges)

    # Fix: use _k to silence warning about unused variable
    indegree = Enum.reduce(edges, %{}, fn {_k, vs}, acc ->
      Enum.reduce(vs, acc, fn v, acc2 -> Map.update(acc2, v, 1, &(&1 + 1)) end)
    end)

    queue = Enum.filter(all_nodes, fn k -> Map.get(indegree, k, 0) == 0 end)
    topo_sort(queue, edges, indegree, "")
  end

  defp topo_sort([], _edges, _indegree, result), do: result

  defp topo_sort(queue, edges, indegree, result) do
    [node | rest] = Enum.sort(queue)
    result = result <> node

    {indegree, new_queue} =
      Enum.reduce(Map.get(edges, node, MapSet.new()), {indegree, rest}, fn neighbor, {ind, q} ->
        updated = Map.update!(ind, neighbor, &(&1 - 1))

        if updated[neighbor] == 0 do
          {updated, [neighbor | q]}
        else
          {updated, q}
        end
      end)

    topo_sort(new_queue, edges, indegree, result)
  end
end

# Usage:
attempts = [
  "319", "680", "180", "690", "129", "620", "762", "689",
  "762", "318", "368", "710", "720", "710", "629", "168",
  "160", "689", "716", "731", "736", "729", "316", "729",
  "729", "710", "769", "290", "719", "680", "318", "389",
  "162", "289", "162", "718", "729", "319", "790", "680",
  "890", "362", "319", "760", "316", "729", "380", "319",
  "728", "716"
]

IO.puts("Derived passcode: #{Euler79.derive_passcode(attempts)}")

